pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/ReplicaGroupInstanceSelector.java

```diff
@@ -152,21 +154,28 @@ private Pair<Map<String, String>, Map<String, String>> selectServersUsingAdaptiv
       SegmentInstanceCandidate selectedInstance = candidates.get(instanceIdx);
       // Adaptive Server Selection
       // TODO: Support numReplicaGroupsToQuery with Adaptive Server Selection.
-      if (!serverRankList.isEmpty()) {
-        int minIdx = Integer.MAX_VALUE;
+      if (!serverRankMap.isEmpty()) {
+        int bestRank = Integer.MAX_VALUE;
         for (SegmentInstanceCandidate candidate : candidates) {
-          int idx = serverRankList.indexOf(candidate.getInstance());
```

Comment:
Thanks for applying this selectedServers optimization as discussed. I think a possibly faster approach is to do the set intersection outside the candidates  loop? smth like 
```
SegmentInstanceCandidate usedCandidate = candidates.stream()
    .filter(c -> selectedServers.contains(c.getInstance()))
    .findFirst()
    .orElse(null);
```

Reply:
The set doesn't make as discussed since we have too check each server of a the segment. Removing it.

Adding the stream for readability, Thanks!